### DISCUSSION TITLE
Spin-Polarise and Analysis

### DISCUSSION THREAD

**Discussion by quantumable** [TRIQS/dft_tools#182] [Q&A]
Hello to all,

I finished the NiO. tutorial and also do the dmft calculation for my system to analyze the 5 orbitals(dxy,dyz,dz2,dxz,dx2y2). 

1-) regarding the analysis, I can't do the representation of imaginary part of Matsubara self energies. here is attached the maxent script used. can you help me to see this quantity?

2-) I also have a system to analyze whose fundamental state is Anti-Ferromagnetic. for this type of calculation in DFT, I simply freeze the magnetic moment in the opposite direction with vasp. to make the DFT+DMFT calculation I must make a DFT calculation (spin-polarized) and then continue the DMFT calculation as I did for the case with 1 correlated atoms where there is a way to make the AFM calculation? what do you advise?

Best regards,

**Comment by the-hampel** [comment]
1) You want to look at the Matsubara self-energy right? Then you do not need MaxEnt. What is exactly your problem when plotting the Matsubara self-energy from the h5 archive? Step by step: 

1. load the self-energy from your h5 archive 
2. choose the block, orbital you want to plot
3. plot the self-energy by using the triqs oplot interface: oplot(Sigma['up'][0,0]) 

which step is not working for you? 

2) Please check also this issue: https://github.com/TRIQS/dft_tools/issues/155 were someone had a similar question: 

> You are asking for spin-polarized DFT calculations if I understand correctly? However, in general I would suggest to always keep the DFT calculation non-magnetic, and only allow in DMFT for magnetic order. This is much more reasonable as you would expect your main magnetic moments to be created by the localized orbitals you want to treat in DMFT. The problem is that the spin-polarized LDA functional is quite different from the non-magnetic one, and cause serious troubles for the charge self-consistency. 

So I would only do the magnetic ordering in DMFT. For this you must remove the spin-degeneracy in SumK and then perform a normal DMFT calculation including two different atomic sites to allow for AFM ordering. I can think about extending the Ni tutorial to the AFM state. But those calculations are quite advanced.

**Comment by quantumable** [comment]
Dear Alex,

1-) step 2 and 3 I did not understand or see an example in the documentation. 
2-) >  For this you must remove the spin-degeneracy in SumK 
   please how to remove the spin-degenerancy in the dmft.py script?  
I never did the spin-polarized calculation with DMFT you suggested to remove the equivalent up/down orbitals from the deg_shells property in Sumk and tried this: but where to insert it?
SK.deg_shells = []
I looked at the soliDMFT code and did not see any reference to magnetic calculation (VASP[non-magnetic]+triqstools(magnetic). and moreover is it necessary as in DFT to initialize the magnetic moment? since I would have two atoms the order in the output files would be spin    dxyup1 | dxydown1........ for first atoms then   dxyup2 | dxydown2...... for second atoms?

3-)if i understood correctly the documentation on the use of wannier90 is not yet available for the charge-self consistent calculation with VASP. and the SOC calculation is also not yet interfaced with vasp where there is a way to do it?

**Reply by the-hampel** [reply]
I think I pointed you already in the issue to the correct documentation pages. You really should do the triqs tutorials to understand the basics on Green's function manipulation and handling: https://github.com/TRIQS/tutorials/tree/3.0.x/TRIQSTutorialsPython .

For your concrete example loading and plotting should work like this: 
```
from h5 import HDFArchive
from triqs.gf import *
from triqs.plot.mpl_interface import plt,oplot

with HDFArchive('NiO.h5','r') as h5:
    print(h5)
    it = h5['DMFT_results']['iteration_count']
    Sigma = h5['DMFT_results']['Iterations']['Sigma_it'+str(it)]
    
fig, ax = plt.subplots(1,figsize=(12,10))

orb = 0
ax.oplot(Sigma['up'][orb,orb].real, label='Re')
ax.oplot(Sigma['up'][orb,orb].imag, label='Im')

ax.set_xlim(-0.1,10)
```
every step in this script can be looked up in the triqs documentation.

**Reply by the-hampel** [reply]
To lift the spin degeneracy in the NiO example: https://github.com/TRIQS/dft_tools/blob/unstable/doc/tutorials/images_scripts/nio.py add in line 36: 
```
SK.deg_shells = [[]]
```
that's all. Now spin up and down are not treated equivalent anymore.

**Reply by the-hampel** [reply]
Regarding point 3): that is correct. This will become available in future releases.

**Comment by quantumable** [comment]
you told me last time that it would also be important to check the convergence after the calculation: how to do it?

**Reply by the-hampel** [reply]
A good start is to compare the density of `G_loc` before the solver runs with the impurity density after the solver is finished, i.e. compare: 
```
S.G_iw.density()
```
with 
```
S.G_imp << SK.extract_G_loc()[0]
S.G_imp.density()
```
in each DMFT iteration. 

Furthermore, you can plot `G0_iw` (similar to how you plot Sigma) for each iteration and see if it converges.

**Comment by quantumable** [comment]
The error using you suggestion:

> Fontconfig warning: ignoring UTF-8: not a valid region tag
> Traceback (most recent call last):
> File "/triqs/install/lib/python3.8/site-packages/triqs/gf/block_gf.py", line 260, in __getitem__
> g = self.__me_as_dict[key]
> KeyError: 'up'
> During handling of the above exception, another exception occurred:
> Traceback (most recent call last):
> File "sigm.py", line 14, in <module>
> ax.oplot(Sigma['up'][orb,orb].imag, label='Im')
> File "/triqs/install/lib/python3.8/site-packages/triqs/gf/block_gf.py", line 262, in __getitem__
> raise IndexError("bloc index '" + repr(key) + "' incorrect. Possible indices are: "+ repr(self.__indices))
> IndexError: bloc index ''up'' incorrect. Possible indices are: ['up_0', 'down_0']

here is for example the script I used to plot the spectral function.

> from h5 import *
> import matplotlib.pyplot as plt
> import numpy as np
> from triqs.gf import *
> 
> 
> filename = 'MnPc'
> 
> ar = HDFArchive(filename+'.h5','a')
> 
> iteration_offset = ar['DMFT_results']['iteration_count']+1
> w = ar['DMFT_results']['Iterations']['w_it'+str(iteration_offset-1)]
> 
> labels = ['dxy', 'dyz', 'dz2', 'dxz', 'dx2y2']
> colors = ['g','b','r' ,'tan','cyan']
> 
> plt.figure(1,figsize=(5,4))
> for iO, orb in enumerate([0,1,2,3,4]):
>     plt.plot(w,ar['DMFT_results']['Iterations']['G_latt_orb_w_o'+str(orb)+'_it'+str(iteration_offset-1)],label=labels[iO],color=colors[iO])
> plt.legend()
> plt.xlim(-10,10)
> plt.xlabel('w (eV)')
> plt.ylabel('A(w)')
> plt.tight_layout()
> plt.savefig('Mn_Aw.png',format='png')
> 
> plt.show()


2-) Regarding the AFM calculation, I would like to summarize it so that you can confirm whether it is correct or not.

I have three systems:
 i-)**single correlated atom**: for this I do the normal calculation with the script present on NiO tutorial; 
 ii-) system of two correlated atoms but coupled Anti-FerroMagnetic(AFM): for this I continue the calculation by keeping the Vasp non-magnetic calculation and lift the spin degeneracy by adding in the script  ` SK.deg_shells = [[]]`

iii-) system of two correlated atoms but coupled FerroMagnetic(FM):  for this I have to do the same calculation made in **I-)   or ii-)**

is there for the last two cases a way to initialize the magnetic moment? the code will be able to differentiate the case **I-)   and ii-)** ?

Or for the case of the coupling of the two atoms the code will know that it makes an intrinsic calculation that it makes for one an FM calculation and for the other an AFM calculation?


It's true that I understood the philosophy of your triqs code but you don't have any documentation where it's really explained how to manipulate and drive the output files? I could read it and do it all by myself. I followed the tutorial on the manipulation of functions and green [ tutorials repository on GitHub](https://github.com/TRIQS/tutorials)  I followed also the tutorial on function manipulation and green and read the triqstools documentation well. 

Best regards.

**Reply by the-hampel** [reply]
Before we discuss any further, you should really to go through some basic tutorials as I suggested multiple times by now. You are providing the following error message: 
```
    File "/triqs/install/lib/python3.8/site-packages/triqs/gf/block_gf.py", line 262, in getitem
    raise IndexError("bloc index '" + repr(key) + "' incorrect. Possible indices are: "+ repr(self.__indices))
    IndexError: bloc index ''up'' incorrect. Possible indices are: ['up_0', 'down_0']
```
And the error tells you exactly what you have to fix!  
```
IndexError: bloc index ''up'' incorrect. Possible indices are: ['up_0', 'down_0']
```
Tells you that you tried to acces the Gf object block as `up` in line 14 of your script, but the only available block indices are `up_0` and `down_0`. This should tell you exactly how to solve the problem, or not?

**Comment by quantumable** [comment]
I followed this tutorial and I don't always have the elements to answer these questions. Is there another tutorial? 
[](https://github.com/TRIQS/tutorials)

**Reply by the-hampel** [reply]
Each Block Green's function object in triqs is accessed like a dictionary in python: https://triqs.github.io/triqs/unstable/documentation/manual/triqs/gfs/py/full.html?highlight=block%20gf . Python tells you in the error message that 'up' is not a correct index of that dictionary. a correct index would be 'up_0' or 'down_0' . So you have to go ahead and change this in the script for plotting.

**Reply by quantumable** [reply]
it works well with your suggestion. the order in this file is as follows dxy dyz dz2 dxz dx2-y2 ?

**Comment by quantumable** [comment]
1-) Just a simple question: when I have two molecules (two atoms correlated) 
there are two ways to make the projection: i-) `LOCPROJ = 1 2 : d : Pr` et  ii-) `LOCPROJ = 1 : d : Pr   LOCPROJ = 2 : d : Pr`   I tested both ways for the first case, I tried dasn the file plo.cfg the following parameter IONS = 1 2 by making IONS = [1 2] it works well (by making plovasp plo.cfg) but the two atoms normally are not equivalent and afterwards the calculation dmft crash.   I used the second way of projecting, then I defined two shells in the plo.cfg file and the dmft calculation works fine. do you have a comment?



2-) Regarding the AFM calculation, I would like to summarize it so that you can confirm whether it is correct or not.
I have three systems:
i-)single correlated atom: for this I do the normal calculation with the script present on NiO tutorial;
ii-) system of two correlated atoms but coupled Anti-FerroMagnetic(AFM): for this I continue the calculation by keeping the Vasp non-magnetic calculation and lift the spin degeneracy by adding in the script SK.deg_shells = [[]]

iii-) system of two correlated atoms but coupled FerroMagnetic(FM): for this I have to do the same calculation made in I-) or ii-)

is there for the last two cases a way to initialize the magnetic moment? the code will be able to differentiate the case I-) and ii-) ?

Or for the case of the coupling of the two atoms the code will know that it makes an intrinsic calculation that it makes for one an FM calculation and for the other an AFM calculation?

**Reply by quantumable** [reply]
Dear Alex,

please how to get also the imaginary part of  hybridization function?

Thanks

**Reply by the-hampel** [reply]
After you run the cthyb solver you can access the hybridization function in the solver object. Suppose your solver is called `S` then `S.Delta_tau` is the hybridization function of the solver. You can then store this object, manipulate, and plot like any other Green's function object.

**Reply by quantumable** [reply]
Dear Alex,

yes indeed I work with three molecule systems. i-) a one molecule system (of which one atom is "correlated") ii-) a two molecule anti-ferromagnetic coupled system (of which two atoms are AFM correlated) iii-) a two molecule ferromagnetic coupled system (of which two atoms are FM correlated).   you suggested me to make the calculation by adding the script dmft this line: SK.deg_shells = [[]] in the AFM case i would like to know if it is also valid for the FM case.


since we are doing an AFM calculation, is there a way to initialize the magnetic moment or the spin in the DMFT calculation? and the analysis will be spin polarized if I understand correctly?

thanks

**Reply by the-hampel** [reply]
This line just lifts the spin degeneracy. The system is thus free to choose any magnetic configuration. To initialize the magnetic moment you have to set the initial self-energies for the solvers in a way that one spin channel is more occupied than the other. 

Note, that I cannot judge how reasonable it is for your system to use DMFT. I am very skeptical here, since you told me that you only have a single atom. Please make sure to understand when DMFT is a useful method for your system at hand.

**Reply by quantumable** [reply]
I think this is a misunderstanding. I have rather three systems of molecules to study separately. one system of a molecule (with a correlated atom) and two systems of two molecules (coupled AFM and FM respectively.)

How to  initialize self-energies for the solvers in a way that one spin channel is more occupied than the other.?

**Reply by the-hampel** [reply]
I understand that, but how many atoms are in each molecule, e.g. how large is each of the molecule?

**Reply by quantumable** [reply]
For the first system, I have 1 molecule adsorbed on the substrate (57 atoms on the molecule with one correlated atom +192 substrate atoms). and the two other systems (114 atoms for the two molecules +192 substrate atoms. the difference is the adsorption configuration whose coupling between the two molecules different)

it is typically this kind of system

<img width="572" alt="Screenshot 2021-10-26 at 16 32 03" src="https://user-images.githubusercontent.com/88863268/138901158-a8efd501-4ec1-422e-a775-61dc823a8362.png">

**Reply by the-hampel** [reply]
Okay that makes sense then. I just wanted to make sure, that it is reasonable to apply DMFT here. 

> how to initialize self-energies for the solvers in a way that one spin channel is more occupied than the other.?

You can first perform a paramagnetic calculation. Then load this self-energy as starting point and modify it. You can add a factor to the up / down channel. Just load the previous results and then do something like: 
```
Sigma_iw['up'] << 1.2*Sigma_iw['up']
Sigma_iw['down'] << 0.8*Sigma_iw['down']
```
This will initialize the magnetic moment for you in the first iteration.

**Reply by quantumable** [reply]
Dear Alex,

can you please confirm? 
When you talk about a paramagnetic calculation, is it a normal calculation like the one in the tutorial?

Here is how I did the calculation please confirm.

1-calculation with the same type of script as in the tutorial.

2-the second calculation I make it in the same direction by adding this 3 lines:

```
SK.deg_shells = [[]]
Sigma_iw['up'] << 1.2*Sigma_iw['up']
Sigma_iw['down'] << 0.8*Sigma_iw['down']

```

what is the role of the factor (1.2 and 0.8) that we add? 

my question was about the initialization of the magnetic moment for the atoms. for example with the vasp-spinpolarized calculation i have for the first atom a magnetic moment of 3.4 and for the second one 3.6 and i wanted to know how to initialize these magnetic moments via this calculation?

Thank you!!

**Reply by the-hampel** [reply]
This procedure is correct. You have to make sure to only do the 1.2*Sigma / 0.8*Sigma once, and not every iteration. That should initialize the magnetic moments. Unless I am missing something it is not straightforward in DMFT to initialize the magnetic moment of the impurity directly. You have to specify a self-energy. And the effect of it on the occupation is not trivial. If you do it like I suggested you can check the spin resolved impurity occupations before the solver starts and then adjust the factor's 1.2 / 0.8 to match your expectations.

**Reply by quantumable** [reply]
Dear Alex, 

I modified the code of this and I got error: 



```
 iteration_offset = mpi.bcast(iteration_offset)
 S.Sigma_iw = mpi.bcast(S.Sigma_iw)
 SK.dc_imp = mpi.bcast(SK.dc_imp)
 SK.dc_energ = mpi.bcast(SK.dc_energ)
 SK.chemical_potential = mpi.bcast(SK.chemical_potential)
```
  Sigma_iw['up'] << 1.2*Sigma_iw['up']
  Sigma_iw['down'] << 0.8*Sigma_iw['down']
```
  # Calc the first G0
 SK.symm_deg_gf(S.Sigma_iw,ish=0)
 SK.put_Sigma(Sigma_imp = [S.Sigma_iw])
 SK.calc_mu(precision=0.01)
 S.G_iw << SK.extract_G_loc()[0]
 SK.symm_deg_gf(S.G_iw, ish=0)
```
 


**error:**

Starting run with 144 MPI rank(s) at : 2021-11-16 00:58:06.531348
Traceback (most recent call last):
  File "/home/5.2/dmft.py", line 118, in <module>
    Sigma_iw['up'] << 1.2*Sigma_iw['up']
NameError: name 'Sigma_iw' is not defined
Traceback (most recent call last):
  File "/home/5.2/dmft.py", line 118, in <module>
    Sigma_iw['up'] << 1.2*Sigma_iw['up']



the error seems to say that I haven't defined Sigma_iw: how to define it?

**Reply by the-hampel** [reply]
Yes indeed, my code was just an example. You have to change the name `Sigma_iw` to the name of your variable that you want to use for the self-energyt. As I said you should start from the paramagnetic solution run before. So Sigma_iw needs to point to that solution. Loaded from an h5 archive.

**Reply by quantumable** [reply]
Dear alex,

Thank you for your answer. I want  to analyze the spin polarized spectral function. 


Indeed as you said I first make a paramagnetic calculation before adding these lines.

I don't understand your explanation. Please could you tell me what I have to add to the script for it to work? or an example?

Thank you.

**Reply by quantumable** [reply]
Dear All,

please do you have an idea to my problem?

Thank you.

**Comment by the-hampel** [comment]
I am confused by your questions. Is your system periodic or do you actually want to calculate a molecule? You are performing DMFT calculations, which strictly speaking are only working reasonably for a lattice problem, hence with a periodic system. Please clarify. 

1) With `IONS = [1 2]` you enforce the sites to be equivalent. It is to be expected that this gives you problems, if the sites are not equivalent.
